File: sunnad-ios/sunnad-ios/Data/Local/Repositories/GroupsLocalRepository.swift
import Foundation

@MainActor
final class GroupsLocalRepository: GroupsRepository {
    private enum Keys {
        static let groups = "sunnad.local.groups.v3"
        static let snapshot = "sunnad.local.groups.snapshot.v3"
    }

    private let userDefaults: UserDefaults
    private let ownerScopeProvider: LocalOwnerScopeProviding?
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init(
        userDefaults: UserDefaults = .standard,
        ownerScopeProvider: LocalOwnerScopeProviding? = nil
    ) {
        self.userDefaults = userDefaults
        self.ownerScopeProvider = ownerScopeProvider
    }

    func fetchGroups() async throws -> [Group] {
        guard let data = userDefaults.data(forKey: storageKey(prefix: Keys.groups)) else {
            return []
        }
        return try decoder.decode([Group].self, from: data)
    }

    func hasCachedSnapshot() -> Bool {
        userDefaults.bool(forKey: storageKey(prefix: Keys.snapshot))
    }

    func replaceGroups(_ groups: [Group]) throws {
        try persist(groups)
    }

    func upsertGroup(_ group: Group) throws {
        var groups = try fetchGroupsSync()
        if let index = groups.firstIndex(where: { $0.id == group.id }) {
            groups[index] = group
        } else {
            groups.append(group)
        }
        try persist(groups)
    }

    func createGroup(name: String) async throws -> Group {
        var groups = try await fetchGroups()
        let me = GroupMember(name: "You", completedToday: 0, totalSharedHabits: 0, sharedHabits: [])
        let group = Group(
            name: name,
            code: String(UUID().uuidString.prefix(8)).uppercased(),
            joinLocked: false,
            members: [me],
            sharedHabitIDs: [],
            ownerMemberID: me.id,
            currentUserMemberID: me.id
        )
        groups.append(group)
        try persist(groups)
        return group
    }

    func joinGroup(code: String) async throws -> Group {
        var groups = try await fetchGroups()
        let me = GroupMember(name: "You", completedToday: 0, totalSharedHabits: 0, sharedHabits: [])
        let group = Group(
            name: "Group \(code.uppercased())",
            code: code.uppercased(),
            joinLocked: false,
            members: [me],
            sharedHabitIDs: [],
            ownerMemberID: me.id,
            currentUserMemberID: me.id
        )
        groups.append(group)
        try persist(groups)
        return group
    }

    func renameGroup(groupID: UUID, name: String) async throws {
        try mutate(groupID: groupID) { group in
            group.name = name
        }
    }

    func setJoinLock(groupID: UUID, locked: Bool) async throws {
        try mutate(groupID: groupID) { group in
            group.joinLocked = locked
        }
    }

    func rotateInviteCode(groupID: UUID) async throws -> String {
        let newCode = String(UUID().uuidString.prefix(8)).uppercased()
        try mutate(groupID: groupID) { group in
            group.code = newCode
        }
        return newCode
    }

    func updateSharing(groupID: UUID, habitIDs: Set<UUID>) async throws {
        try mutate(groupID: groupID) { group in
            group.sharedHabitIDs = habitIDs
        }
    }

    func leaveGroup(groupID: UUID) async throws {
        var groups = try await fetchGroups()
        groups.removeAll { $0.id == groupID }
        try persist(groups)
    }

    func deleteGroup(groupID: UUID) async throws {
        try await leaveGroup(groupID: groupID)
    }

    func kickMember(groupID: UUID, memberUserID: UUID) async throws {
        try mutate(groupID: groupID) { group in
            group.members.removeAll { $0.id == memberUserID }
        }
    }

    func setProgressDisplayMode(groupID: UUID, mode: GroupProgressDisplayMode) async throws {
        try mutate(groupID: groupID) { group in
            group.progressDisplayMode = mode
        }
    }

    func refreshGroup(groupID: UUID) async throws -> Group? {
        try await fetchGroups().first(where: { $0.id == groupID })
    }

    func sendNudge(groupID: UUID, toUserID: UUID, habitID: UUID) async throws -> GroupNudgeStatus {
        _ = (groupID, toUserID, habitID)
        return .recipientNotRegistered
    }

    private func mutate(groupID: UUID, _ mutation: (inout Group) -> Void) throws {
        var groups = try fetchGroupsSync()
        guard let index = groups.firstIndex(where: { $0.id == groupID }) else {
            return
        }
        mutation(&groups[index])
        try persist(groups)
    }

    private func fetchGroupsSync() throws -> [Group] {
        guard let data = userDefaults.data(forKey: storageKey(prefix: Keys.groups)) else {
            return []
        }
        return try decoder.decode([Group].self, from: data)
    }

    private func persist(_ groups: [Group]) throws {
        let data = try encoder.encode(groups)
        userDefaults.set(data, forKey: storageKey(prefix: Keys.groups))
        userDefaults.set(true, forKey: storageKey(prefix: Keys.snapshot))
    }

    private func storageKey(prefix: String) -> String {
        let scope = ownerScopeProvider?.currentOwnerScopeRawValue ?? LocalOwnerScope.guest.rawValue
        return "\(prefix).\(scope)"
    }
}

/// Keeps the last complete server snapshot available while the network is down.
/// Server-owned operations still fail honestly when they require authorization or delivery.
@MainActor
final class CachedGroupsRepository: GroupsRepository {
    private let remote: GroupsRepository
    private let cache: GroupsLocalRepository

    init(remote: GroupsRepository, cache: GroupsLocalRepository) {
        self.remote = remote
        self.cache = cache
    }

    func fetchGroups() async throws -> [Group] {
        do {
            let groups = try await remote.fetchGroups()
            try cache.replaceGroups(groups)
            return groups
        } catch {
            guard cache.hasCachedSnapshot() else { throw error }
            return try await cache.fetchGroups()
        }
    }

    func createGroup(name: String) async throws -> Group {
        let group = try await remote.createGroup(name: name)
        try? cache.upsertGroup(group)
        return group
    }

    func joinGroup(code: String) async throws -> Group {
        let group = try await remote.joinGroup(code: code)
        try? cache.upsertGroup(group)
        return group
    }

    func renameGroup(groupID: UUID, name: String) async throws {
        try await remote.renameGroup(groupID: groupID, name: name)
        try? await cache.renameGroup(groupID: groupID, name: name)
    }

    func setJoinLock(groupID: UUID, locked: Bool) async throws {
        try await remote.setJoinLock(groupID: groupID, locked: locked)
        try? await cache.setJoinLock(groupID: groupID, locked: locked)
    }

    func rotateInviteCode(groupID: UUID) async throws -> String {
        let code = try await remote.rotateInviteCode(groupID: groupID)
        await refreshCachedGroup(groupID: groupID)
        return code
    }

    func updateSharing(groupID: UUID, habitIDs: Set<UUID>) async throws {
        try await remote.updateSharing(groupID: groupID, habitIDs: habitIDs)
        try? await cache.updateSharing(groupID: groupID, habitIDs: habitIDs)
    }

    func leaveGroup(groupID: UUID) async throws {
        try await remote.leaveGroup(groupID: groupID)
        try? await cache.leaveGroup(groupID: groupID)
    }

    func deleteGroup(groupID: UUID) async throws {
        try await remote.deleteGroup(groupID: groupID)
        try? await cache.deleteGroup(groupID: groupID)
    }

    func kickMember(groupID: UUID, memberUserID: UUID) async throws {
        try await remote.kickMember(groupID: groupID, memberUserID: memberUserID)
        try? await cache.kickMember(groupID: groupID, memberUserID: memberUserID)
    }

    func setProgressDisplayMode(groupID: UUID, mode: GroupProgressDisplayMode) async throws {
        try await remote.setProgressDisplayMode(groupID: groupID, mode: mode)
        try? await cache.setProgressDisplayMode(groupID: groupID, mode: mode)
    }

    func refreshGroup(groupID: UUID) async throws -> Group? {
        do {
            let group = try await remote.refreshGroup(groupID: groupID)
            if let group {
                try? cache.upsertGroup(group)
            }
            return group
        } catch {
            guard cache.hasCachedSnapshot() else { throw error }
            return try await cache.refreshGroup(groupID: groupID)
        }
    }

    func sendNudge(groupID: UUID, toUserID: UUID, habitID: UUID) async throws -> GroupNudgeStatus {
        try await remote.sendNudge(groupID: groupID, toUserID: toUserID, habitID: habitID)
    }

    private func refreshCachedGroup(groupID: UUID) async {
        guard let group = try? await remote.refreshGroup(groupID: groupID) else { return }
        try? cache.upsertGroup(group)
    }
}

File: sunnad-android/app/src/main/java/com/arystan/almasuly/sunnadandroid/data/local/repository/GroupsLocalRepository.kt
package com.arystan.almasuly.sunnadandroid.data.local.repository

import com.arystan.almasuly.sunnadandroid.core.local.OwnerScopeResolver
import com.arystan.almasuly.sunnadandroid.core.model.Group
import com.arystan.almasuly.sunnadandroid.core.model.GroupMember
import com.arystan.almasuly.sunnadandroid.core.model.GroupNudgeStatus
import com.arystan.almasuly.sunnadandroid.core.model.GroupProgressDisplayMode
import com.arystan.almasuly.sunnadandroid.domain.repository.GroupsRepository
import com.arystan.almasuly.sunnadandroid.data.local.dao.GroupDao
import com.arystan.almasuly.sunnadandroid.data.local.mappers.toDomain
import com.arystan.almasuly.sunnadandroid.data.local.mappers.toEntity
import com.arystan.almasuly.sunnadandroid.data.local.mappers.toMemberEntities
import com.arystan.almasuly.sunnadandroid.data.local.mappers.toSharedHabitEntities
import java.util.UUID

class GroupsLocalRepository(
    private val groupDao: GroupDao,
    private val ownerScopeResolver: OwnerScopeResolver
) : GroupsRepository {
    override suspend fun fetchGroups(): List<Group> {
        val scope = ownerScopeResolver.currentOwnerScopeRaw
        val groups = groupDao.fetchGroups(scope)
        return groups.map { group ->
            val members = groupDao.fetchMembers(group.id)
            val shared = groupDao.fetchSharedHabits(group.id)
            group.toDomain(members, shared)
        }
    }

    override suspend fun createGroup(name: String): Group {
        val me = GroupMember(name = "You", completedToday = 0, totalSharedHabits = 0, sharedHabits = emptyList())
        val group = Group(
            name = name,
            code = UUID.randomUUID().toString().take(8).uppercase(),
            members = listOf(me),
            sharedHabitIds = emptySet(),
            ownerMemberId = me.id,
            currentUserMemberId = me.id
        )
        persistGroup(group)
        return group
    }

    override suspend fun joinGroup(code: String): Group {
        val me = GroupMember(name = "You", completedToday = 0, totalSharedHabits = 0, sharedHabits = emptyList())
        val group = Group(
            name = "Group ${code.uppercase()}",
            code = code.uppercase(),
            members = listOf(me),
            sharedHabitIds = emptySet(),
            ownerMemberId = me.id,
            currentUserMemberId = me.id
        )
        persistGroup(group)
        return group
    }

    override suspend fun renameGroup(groupId: UUID, name: String) {
        mutate(groupId) { it.copy(name = name) }
    }

    override suspend fun setJoinLock(groupId: UUID, locked: Boolean) {
        mutate(groupId) { it.copy(joinLocked = locked) }
    }

    override suspend fun rotateInviteCode(groupId: UUID): String {
        val next = UUID.randomUUID().toString().take(8).uppercase()
        mutate(groupId) { it.copy(code = next) }
        return next
    }

    override suspend fun updateSharing(groupId: UUID, habitIds: Set<UUID>) {
        mutate(groupId) { it.copy(sharedHabitIds = habitIds) }
    }

    override suspend fun leaveGroup(groupId: UUID) {
        val scope = ownerScopeResolver.currentOwnerScopeRaw
        groupDao.deleteGroup(scope, groupId.toString())
        groupDao.deleteMembers(groupId.toString())
        groupDao.deleteSharedHabits(groupId.toString())
    }

    override suspend fun deleteGroup(groupId: UUID) {
        leaveGroup(groupId)
    }

    override suspend fun kickMember(groupId: UUID, memberUserId: UUID) {
        mutate(groupId) { group ->
            group.copy(members = group.members.filterNot { it.id == memberUserId })
        }
    }

    override suspend fun setProgressDisplayMode(groupId: UUID, mode: GroupProgressDisplayMode) {
        mutate(groupId) { group ->
            group.copy(progressDisplayMode = mode)
        }
    }

    override suspend fun refreshGroup(groupId: UUID): Group? {
        return fetchGroups().firstOrNull { it.id == groupId }
    }

    override suspend fun sendNudge(groupId: UUID, toUserId: UUID, habitId: UUID): GroupNudgeStatus {
        Triple(groupId, toUserId, habitId)
        return GroupNudgeStatus.SENT
    }

    private suspend fun mutate(groupId: UUID, update: (Group) -> Group) {
        val existing = refreshGroup(groupId) ?: return
        persistGroup(update(existing))
    }

    private suspend fun persistGroup(group: Group) {
        val now = System.currentTimeMillis()
        val scope = ownerScopeResolver.currentOwnerScopeRaw
        groupDao.replaceGroupSnapshot(
            group = group.toEntity(scope, now),
            members = group.toMemberEntities(now),
            sharedHabits = group.toSharedHabitEntities(now)
        )
    }
}
